src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -398,7 +398,13 @@ internal bool CheckValueKind(SyntaxNode node, BoundExpression expr, BindValueKin
 
                     // SPEC: When this is used in a primary-expression within an instance method or instance accessor
                     // SPEC: of a struct, it is classified as a variable. 
-                    if (!thisref.Type.IsValueType || RequiresReturnableReference(valueKind))
+
+                    // Note: RValueOnly is checked at the beginning of this method. Since we are here we need more than readable.
+                    //"this" is only readable in members of readonly structs, unless we are in a constructor.
+                    //"this" is not returnable by reference in a struct.
+                    if (!thisref.Type.IsValueType ||
+                        (thisref.Type.IsReadOnly && (this.ContainingMember() as MethodSymbol)?.MethodKind != MethodKind.Constructor) ||
```

Comment:
Well, turns out "this" is not accessible at all inside nested methods in structs. Not even for reading or any other kind of use.
I will change to ContainingMemberOrLambda, since it seems to be the right thing to do, and add related tests, but the difference would not observable right now.

---
In reply to: [118796219](https://github.com/dotnet/roslyn/pull/19713#discussion_r118796219) [](ancestors = 118796219,118784631)